numpy/fft/_pocketfft_umath.cpp

```diff
@@ -0,0 +1,437 @@
+        PyErr_NoMemory();
+    }
+    catch (const std::exception& e) {
+        PyErr_SetString(PyExc_RuntimeError, e.what());
+    }
+    NPY_DISABLE_C_API;
+}
+
+
+/*
+ * Transfer to and from a contiguous buffer.
+ * copy_input: copy min(nin, n) elements from input to buffer and zero rest.
+ * copy_output: copy n elements from buffer to output.
+ */
+template <typename T>
+static inline void
+copy_input(char *in, npy_intp step_in, size_t nin,
+           T buff[], size_t n)
+{
+    size_t ncopy = nin <= n ? nin : n;
+    char *ip = in;
+    size_t i;
+    for (i = 0; i < ncopy; i++, ip += step_in) {
+      buff[i] = *(T *)ip;
+    }
+    for (; i < n; i++) {
+      buff[i] = 0;
+    }
+}
+
+template <typename T>
+static inline void
+copy_output(T buff[], char *out, npy_intp step_out, size_t n)
+{
+    char *op = out;
+    for (size_t i = 0; i < n; i++, op += step_out) {
+        *(T *)op = buff[i];
+    }
+}
+
+/*
+ * Gufunc loops calling the pocketfft code.
+ */
+template <typename T>
+static void
+fft_loop(char **args, npy_intp const *dimensions, ptrdiff_t const *steps,
+         void *func)
+{
+    char *ip = args[0], *fp = args[1], *op = args[2];
+    size_t n_outer = (size_t)dimensions[0];
+    ptrdiff_t si = steps[0], sf = steps[1], so = steps[2];
+    size_t nin = (size_t)dimensions[1], nout = (size_t)dimensions[2];
+    ptrdiff_t step_in = steps[3], step_out = steps[4];
+    bool direction = *((bool *)func); /* pocketfft::FORWARD or BACKWARD */
+
+    TRY_FOR_CPP_EXCEPTIONS;
+    assert (nout > 0);
+
+#ifndef POCKETFFT_NO_VECTORS
+    /*
+     * For the common case of nin >= nout, fixed factor, and suitably sized
+     * outer loop, we call pocketfft directly to benefit from its vectorization.
+     * (For nin>nout, this just removes the extra input points, as required;
+     * the vlen constraint avoids compiling extra code for longdouble, which
+     * cannot be vectorized so does not benefit.)
+     */
+    constexpr auto vlen = pocketfft::detail::VLEN<T>::val;
+    if (vlen > 1 && n_outer >= vlen && nin >= nout && sf == 0) {
+        std::vector<size_t> shape = { n_outer, nout };
+        std::vector<ptrdiff_t> strides_in = { si, step_in };
+        std::vector<ptrdiff_t> strides_out = { so, step_out};
+        std::vector<size_t> axes = { 1 };
+        pocketfft::c2c(shape, strides_in, strides_out, axes, direction,
+                       (std::complex<T> *)ip, (std::complex<T> *)op, *(T *)fp);
+        return;
+    }
+#endif
+    /*
+     * Otherwise, use a non-vectorized loop in which we try to minimize copies.
+     * We do still need a buffer if the output is not contiguous.
+     */
+    auto plan = pocketfft::detail::get_plan<pocketfft::detail::pocketfft_c<T>>(nout);
+    void *buff = NULL;
+    if (step_out != sizeof(std::complex<T>)) {
+        buff = pocketfft::detail::aligned_alloc(sizeof(std::complex<T>),
+                                                nout * sizeof(std::complex<T>));
```

Comment:
There is a small problem here that `buff` is a `void *` and not a `pocketfft:detail::arr<std::complex<T>>`.  Which I think means that if the `plan->exec` fails below no cleanup will occur.
My C++ is a bit trial and error, but I think:
```
/* Create a temporary array to hold a buffer if op is not contiguous */
bool contiguous_op = (step_out != sizeof(std::complex<T>));
pocketfft::detail::arr<std::complex<T>> tmp_arr(contiguous_op ? 0 : nout);
```
should be an option (an array of length 0 doesn't do a malloc so that is fine).

And below:
```
std::complex<T> *op_or_buff = (contiguous_op ? (std::complex<T> *)op : tmp_arr.data());
```

(Clearly, the dealloc below then becomes unnecessary, which is the whole point.)

Reply:
OK, that makes a lot of sense! I implemented it (though with an opposite bool, `buffered`).